### ISSUE TITLE
route-less engine "crashes" on component property mutation

### ISSUE THREAD

**Issue by thousand** [dgeb/ember-engines#126]
I'm developing a route-less engine. It has a main parent component that handles opening and closing some child components. That main parent component keeps the data for its children in an array property and shows one child for each element in the array. The biggest issue i discovered around this is that if I mutate that array, adding or removing an element causing a child component to be added or removed, the engine gets removed from the DOM entirely, with no warnings or errors in the console. Additionally the component lifecycle event handlers like `willUpdate` or `willDestroyElement` are called. Example code:

```js
// parent-component.js
import Ember from 'ember';
export default Ember.Component.extend({
  init() {
    this.openChildren = [{name: 'John'}, {name: 'Paul'}, {name: 'George'}]
  },

  willUpdate() {
    // i'm never called
    console.log('updating...');
  },

  willDestroyElement() {
    // i'm never called
    console.log('destroying...');
  },

  actions: {
    openChild(child = {name: "Ringo"}) {
      let openChildren = this.get('openChildren');
      openChildren.addObject(child);
    },

    closeChild(child) {
      let openChildren = this.get('openChildren');
      openChildren.removeObject(conversation);
    }
  }

});
```

```hbs
{! parent-component.hbs }
<h1>Parent</h1>

{{#each openChildren as |child|}}
  {{child
    child=child
    closeSelectedChild="closeChild"
  }}
{{/each}}
```

```js
// child-component.js
import Ember from 'ember';
export default Ember.Component.extend({
  actions: {
    close() {
      this.sendAction('closeSelectedChild', this.get('child'));
    }
  }
});
```

```hbs
{! child-component.hbs }
<header>
  <h2>{{name}}</h2>
  <button {{action="close"}}>X</button>
</header>
```

Simpler component property mutations are also not working, though without the catastrophic results. Just a simple iterator like `<button {{action "increment"}}>{{value}}</button>` also does not update, and throws this error in the console on the first mutation: `Uncaught Error: Assertion Failed: A helper named 'parent-component' could not be found`

**Comment by trentmwillis** [comment]
I feel like the root issue here is probably similar to the one in #85.

@thousand could you put together a simple app to clone down so we can debug further? Or, even better, would be if you could set up a failing test in the dummy app inside this addon.

**Comment by thousand** [comment]
@trentmwillis Sorry for the delay on this. I've not been able to create a simple test app with an engine. Following the instructions in the docs for `ember-engines`, i get an dependency resolution error for the `ember-metal/streams/utils` dependency in the `mount` keyword definition. Copying the ember version from my currently working app and engine, `"ember": "kratiahuja/ember-preparse-patch#5d2926e"` at least solves that dependency issue, but now my engine is not found when using `{{mount}}` for the routeless engine. I haven't tried a routed version yet, but will soon.

Current state is living here: [https://github.com/thousand/engine-test](https://github.com/thousand/engine-test)

**Comment by thousand** [comment]
Ok, i was able to resolve my issues with getting a basic engine running. After much head scratching I discovered you need to use the ember-engines resolver as the app's resolver as well >.<. [My engine-test repo from before](https://github.com/thousand/engine-test) now contains a "working" app and engine. Additionally, after looking through some of the issues here, discovered that ember-engines works with the ember beta channel, so the repo has been updated to reflect that more ideal state. However, the engine is now not rendering components at all.

**Comment by thousand** [comment]
Ok, now that i've thoroughly established that i'm an idiot, [my test repo](https://github.com/thousand/engine-test) now demonstrates the behavior.

**Comment by dgeb** [comment]
@thousand could you please update your test repo to use ember-engines 0.2.5 and ember 2.6.0 (final) and verify that your problem is still present. If it is, please describe the steps to reproduce or add a failing test and I will investigate.

**Comment by thousand** [comment]
@dgeb versions have been updated as suggested. The issue is still there, but slightly different from my original description, it indeed looks like the same issue that @trentmwillis noted in #85 and fixed in #92. When a property changes that causes the component to re-render, the component (not the engine, as i improperly assumed before) disappears on re-render. I didn't get around to writing a test, but in my example, clicking any of the buttons there will reproduce it. There's two examples in there. Each is a component called directly from the application.hbs template of the engine. One mutates an array, and the other is a simple counter with a button to increment. I did (briefly) try to port @trentmwillis's fix in PR #92 to the `mount` keyword but haven't gotten it working yet. I'll continue looking at it as well, now that i understand a little better about how ember is resolving and rendering components.

**Comment by thousand** [comment]
@dgeb @trentmwillis  Ok, here's the results of my investigations thus far.

I'm not sure WHY exactly this is a problem, but i'm pretty sure the `renderEnv` being generated for components (and other render nodes) used in the engine is wrong as those nodes re-render. In the routed engines, the `engineInstance`s are kept around inside the route setup. In the routeless engines, they're created inside the `mount` keyword's `render` method, but not stored anywhere to be used on subsequent child nodes' renders. When the child nodes of the engine are re-rendered and `mount`'s `childEnv` method is called to build render `env` objects, it cannot regain a handle on the `engineInstance` and more importantly the engine's toplevel/application view. This leaves no way to build the proper `env` for child nodes of the engine. Unless i've really missed something, this appears to be an existential problem with routeless engines, or at least with the current implementation where the `mount` keyword's `render` method spins up the `engineInstance` without stashing it anywhere.
